### PR TITLE
All pages in place view

### PR DESCRIPTION
**Pull request by portuiu**

### CODE REVIEW

**Comment by dbarashev** [comment]
Hey.
You have write access to papeeria-pdfjs repository. Please create a new branch forked from papeeria-pdfjs:master, push it to papeeria-pdfjs repository and create a pull request to master.

File: workspace/viewer/PdfJsViewer.ts [hunk 1]
```diff
@@ -310,6 +310,7 @@ export class PdfJsViewer {
               private readonly logger: Logger,
               private readonly utils: Utils,
               private readonly i18n: I18N) {
+    this.pdfPageView = [];
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: Please change the type and the name of this member variable, because it is not a single PDFPageView anymore.

File: workspace/viewer/PdfJsViewer.ts [hunk 2]
```diff
@@ -364,6 +362,14 @@ export class PdfJsViewer {
     }
   }
 
+  public showAll(url:string, isResize: boolean = false){
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: nit: insert space after `url:`

File: workspace/viewer/PdfJsViewer.ts [hunk 3]
```diff
@@ -364,6 +362,14 @@ export class PdfJsViewer {
...
+    var pdfViewer = this;
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: In TypeScript you don't need these tricks. Use [fat arrows](https://basarat.gitbooks.io/typescript/docs/arrow-functions.html) to bind callback functions to proper `this`

File: workspace/viewer/PdfJsViewer.ts [hunk 4]
```diff
@@ -364,6 +362,14 @@ export class PdfJsViewer {
...
+    pdfjs.getDocument(url).then(function(pdf){
+      for(var i = 1; i <= pdf.numPages; i++){
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: In TypeScript world people use `let` instead of `var`. Also, please insert a space after `for`

File: workspace/viewer/PdfJsViewer.ts [hunk 5]
```diff
@@ -387,10 +393,9 @@ export class PdfJsViewer {
       const onDocumentSuccess = (pdf) => {
         this.currentFile = pdf;
         this.currentFileUrl = task.url;
-        if (this.currentPage > pdf.numPages) {
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: Why did you remove this? `currentPage` is supplied externally and it *really* can be greater than the number of pages. And no matter if we have smooth scrolling or not, we'll want to show user that very `currentPage`.

File: workspace/viewer/PdfJsViewer.ts [hunk 6]
```diff
@@ -444,8 +448,10 @@ export class PdfJsViewer {
     return true
   }
 
-  private positionCanvas() {
-    const canvas = $(".canvasWrapper", this.jqRoot);
+  private positionCanvas(pageNumber: number) {
+    const parent = $("#pageContainer" + pageNumber, this.jqRoot);
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: There is nifty [string interpolation](https://basarat.gitbooks.io/typescript/docs/template-strings.html) in TypeScript

File: workspace/viewer/PdfJsViewer.ts [hunk 7]
```diff
@@ -464,18 +470,20 @@ export class PdfJsViewer {
...
       top : canvasOffset.top,
       left : canvasOffset.left
     });
   }
 
   resetCanvas() {
-    if (this.pdfPageView) {
-      this.pdfPageView.destroy();
+    if (this.pdfPageView.length != 0) {
+      for(var i = 0; i != this.pdfPageView.length; i++){
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: [Make loops great again](https://basarat.gitbooks.io/typescript/docs/for...of.html). `for (let page of this.pdfPageView)`

File: workspace/viewer/PdfJsViewer.ts [hunk 8]
```diff
@@ -505,19 +513,6 @@ export class PdfJsViewer {
     }
   }
 
-  pageUp = () => {
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: We need these functions in production, please get them back and make them working.

**Comment by dbarashev** [comment]
This pull request looks like a good start (after the issues will have been fixed), however, it can't yet be merged into production (see below what we can do with this).

So this change will basically add as many canvas elements as we have pages in the document, and load all pdf pages. The former looks okay, but the latter will make user experience worse for big documents. 

What we want is to load page on demand if it shows up when scrolling. It seems that on-demand loading will be a natural evolution after this change.

With respect to branches and merging, it can't be merged into `master` (assuming that `master` here means "what we have in production"), but it worth merging somewhere just to keep pull requests small.  I suggest merging it into a separate "project" branch. Please create a branch called `project-smooth-scrolling` in `bardsoftware/papeeria-pdfjs` repo and target these changes at that branch (that is, merge request from `your_change_branch` into `project-smooth-scrolling`)

**Comment by portuiu** [comment]
Hi! Thank you for helpful review.

I made some changes, but in the same branch, because I dont have permissions to create new one on `bardsoftware/papeeria-pdfjs`.

File: workspace/viewer/PdfJsViewer.ts (hunk 2 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts (hunk 3 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts (hunk 4 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts (hunk 6 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts (hunk 7 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts (hunk 8 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts (hunk 1 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: I changed the name pdfPageView to pdfPagesView, but didn't figure out yet how to create an array of PDFPageView.PDFPageView objects, because it throws TS2503 error "Cannot find namespace PDFPageView".

File: workspace/viewer/PdfJsViewer.ts (hunk 5 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: This part of code loses relevance if we want to load and show more than one document's page, because it will refer to last loaded page. So I moved this logic to another place.

Comment (dbarashev) [reply workspace/viewer/PdfJsViewer.ts]: Can you explain this? 

> because it will refer to last loaded page

`currentPage` is not the last loaded, it is the page user had been looking at until we reloaded the document. If user had been looking at page 10 and in the new PDF we have only 9 pages, we have to set `currentPage` properly.

File: workspace/viewer/PdfJsViewer.ts (hunk 1 above)
Comment (dbarashev) [reply workspace/viewer/PdfJsViewer.ts]: Perhaps TypeScript type definitions in `node_modules` diverged from the real js code (we have outdated pdfjs library in this repository).
Anyway, the edit distance between `pdfPageView` and `pdfPagesView` is next to nothing while conceptually these objects are "two big differences", and array  is not a view at all. Rename it to `allPages` or `loadedPages` or anything else which matches the semantics.

**Review by dbarashev — COMMENTED** [review]
There are many done's here, but after all these commits and reverts there are no real changes comparing to the first pullrequested version. Ant it seems that pr #3 is identical to this one, modulo compiled JS code. What shall I review?

File: workspace/viewer/PdfJsViewer.ts (hunk 5 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: From this point, you're right. Because I wanted to update currentPage from scroll event, so loading new PDF would reset it and this part of code would look completely useless.
Returned it back.

File: workspace/viewer/PdfJsViewer.ts [hunk 9]
```diff
@@ -288,7 +288,8 @@ export class PdfJsViewer {
   static getPresetScales(): number[] { return Zoom.ZOOM_FACTORS; }
 
   // These are from pdfjs library
-  pdfPageView?: PDFPageViewStatic;
+  loadedPages?: PDFPageViewStatic[] = [];
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: It appears that it is never `undefined` so you can remove `?` from the name (`?` is a signal to compiler that variable can be undefined, without `?` we are certain that it is defined always)

File: workspace/viewer/PdfJsViewer.ts [hunk 10]
```diff
@@ -288,7 +288,8 @@ export class PdfJsViewer {
...
+  numPages: number;
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: Where is it assigned?

File: workspace/viewer/PdfJsViewer.ts [hunk 11]
```diff
@@ -394,14 +399,13 @@ export class PdfJsViewer {
...
       this.currentTask = task;
-      this.currentPage = task.page;
       const onDocumentSuccess = (pdf) => {
         this.currentFile = pdf;
         this.currentFileUrl = task.url;
         if (this.currentPage > pdf.numPages) {
           this.currentPage = pdf.numPages;
         }
-        this.openPage(pdf, this.currentPage);
+        this.openPage(pdf, this.currentTask.page);
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]:
```
workspace/viewer/PdfJsViewer.ts(408,28): error TS2532: Object is possibly 'undefined'.
```

`currentTask` is allowed to be undefined and at this moment it can become undefined or assigned some other value. If you meant to use `page` from `task` above, please use `task.page`. If you really meant to use `page` from `currentTask`, whatever it is, you need to check if `currentTask` is still defined.

File: workspace/viewer/PdfJsViewer.ts [hunk 12]
```diff
@@ -478,29 +480,32 @@ export class PdfJsViewer {
...
     textLayer.css({
       top : canvasOffset.top,
       left : canvasOffset.left
     });
   }
 
   resetCanvas() {
-    if (this.pdfPageView) {
-      this.pdfPageView.destroy();
+    if (this.loadedPages.length != 0) {
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: What's the purpose of this `if`? If array is empty we will just make no loops below without additional ifs.

File: workspace/viewer/PdfJsViewer.ts [hunk 13]
```diff
@@ -478,29 +480,32 @@ export class PdfJsViewer {
...
+      for(let page of this.loadedPages){
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: nit: pls add a space after `for`

File: workspace/viewer/PdfJsViewer.ts [hunk 14]
```diff
@@ -478,29 +480,32 @@ export class PdfJsViewer {
...
     this.queue.clear();
   }
 
-  private openCurrentPage() {
-    if (this.currentFileUrl) {
-      let lastCompletedMainFileId = (this.queue.isEmpty() && this.queue.lastCompleted)
-          ? this.queue.lastCompleted.mainFileId : undefined;
-      this.show(this.currentFileUrl, this.currentPage, true, lastCompletedMainFileId);
+  public showPage(page: number) {
+    if(page > this.numPages){
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: nit: space after `if`

File: workspace/viewer/PdfJsViewer.ts [hunk 15]
```diff
@@ -478,29 +480,32 @@ export class PdfJsViewer {
...
+      page = this.numPages;
     }
+    let pageWrapper = $("div").find(`[data-page-number='${page}']`)[0];
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: You just search in all divs on the page? This looks like pretty broad target. Also, is anyone defining those `data-page-number` attrs?

File: workspace/viewer/PdfJsViewer.ts [hunk 16]
```diff
@@ -515,49 +520,51 @@ export class PdfJsViewer {
...
 
   zoomOut = () => {
-    this.zoom.zoomOut() && this.openCurrentPage();
+    this.zoom.zoomOut() && this.showPage(this.currentPage);
   };
 
   zoomWidth = () => {
     this.zoom.setFitting(ZoomingMode.FIT_WIDTH);
-    this.openCurrentPage();
+      this.showPage(this.currentPage);
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: Something is wrong with the indents here.

File: workspace/viewer/PdfJsViewer.ts [hunk 17]
```diff
@@ -515,49 +520,51 @@ export class PdfJsViewer {
...
     }
   }
 
   pageUp = () => {
-    if (this.currentPage > 1) {
-      this.currentPage -= 1;
-      this.openCurrentPage();
-    }
+      if (this.currentPage > 1) {
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: Something is wrong with the indents. We use 2 spaces.

File: workspace/viewer/PdfJsViewer.ts (hunk 9 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Thanks for information. Read the documentation once again and fixed it.

File: workspace/viewer/PdfJsViewer.ts (hunk 11 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts (hunk 12 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts (hunk 13 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts (hunk 14 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts [hunk 18]
```diff
@@ -478,29 +480,30 @@ export class PdfJsViewer {
...
       left : canvasOffset.left
     });
   }
 
   resetCanvas() {
-    if (this.pdfPageView) {
-      this.pdfPageView.destroy();
-    }
-    this.pdfPageView = undefined;
+	  for (let page of this.loadedPages){
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: Please adjust your IDE settings. No tabs. 2 spaces indents.

File: workspace/viewer/PdfJsViewer.ts (hunk 15 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts (hunk 17 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts (hunk 16 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts [hunk 19]
```diff
@@ -478,29 +480,30 @@ export class PdfJsViewer {
...
-  private openCurrentPage() {
-    if (this.currentFileUrl) {
-      let lastCompletedMainFileId = (this.queue.isEmpty() && this.queue.lastCompleted)
-          ? this.queue.lastCompleted.mainFileId : undefined;
-      this.show(this.currentFileUrl, this.currentPage, true, lastCompletedMainFileId);
+  public showPage(page: number) {
+    if (page > this.numPages){
+      page = this.numPages;
     }
+    let pageWrapper = this.loadedPages.find(x => x.id === page).div;
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: Here it makes sense to use thin arrow because you don't need `this` in the filter: `x  -> x.id === page`. By the way, I see that `id` is defined by our own call but what about `div`? Is it defined somewhere in the library and means what dowe think it means?

File: workspace/viewer/PdfJsViewer.ts (hunk 10 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Not fixed yet.

Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Why do I should create Interface to `Array<PDFPageViewStatic>` if `Array<T>` is implemented in ES5 by default?

Comment (dbarashev) [reply workspace/viewer/PdfJsViewer.ts]: Sorry? Is this question related to numPages or what? And what do you mean when you say "create interface"? Here (at line above actually) we just declare a variable of typed array, we do not create any interfaces. It is probably compiled to Array<PDFPageViewStatic> in ES5, but we write in TypeScript.

File: workspace/viewer/PdfJsViewer.ts (hunk 19 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Thank you for the advice. 
Yes, it's defined by library. `div` is the variable of `PDFPageView` object. It stores `div`, which contains view of page.
`id` is also variable of this object like `div`.

Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: I haven't found any information about thin arrow functions neither in TypeScript nor in ES5/6. They exist only in CoffeeScript, but can be added in future.

File: workspace/viewer/PdfJsViewer.ts (hunk 18 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done. It's caused because I've swaped my IDE to Vim recently.)

File: workspace/viewer/PdfJsViewer.ts (hunk 10 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Sorry, this question wasn't related to numPages. 
Fixed.

File: workspace/viewer/PdfJsViewer.ts (hunk 19 above)
Comment (dbarashev) [reply workspace/viewer/PdfJsViewer.ts]: Oh yup, sorry. I thought they were in TS as well. We can replace them with just `function` but this will probably make this "lambda" way more wordy, so let's leave as is. 
However, you need to do something when `find` returns `undefined` which seems to be possible. For instance, `resetCanvas` call resets the array, but leaves `currentfile` as is, so subsequent call to `showPage` will crash.

File: workspace/viewer/PdfJsViewer.ts [hunk 20]
```diff
@@ -13,11 +13,12 @@ interface PDFPageViewStatic {
   new(options: any) : PDFPageViewStatic;
   destroy(): void;
 
-  setPdfPage(pdfPage: PDF.PDFPageProxy): void;
+  setPdfPage(pdfPage:  PDF.PDFPageProxy): void;
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: why this space?

Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts (hunk 19 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Yes, you are right. We associate our canvas with one pdf file, so resetting the canvas should also set `currentfile` to `undefined`.
Done.

File: workspace/viewer/PdfJsViewer.ts (hunk 20 above)
Comment (dbarashev) [reply workspace/viewer/PdfJsViewer.ts]: I meant the space symbol after `pdfPage:`. There are two instead of one, the extra space was added by you in [this commit](https://github.com/bardsoftware/papeeria-pdfjs/pull/1/commits/5ba7e3dd2c135a3081ee1ef92d890f4ba4e3ddf7)

Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts [hunk 21]
```diff
@@ -478,29 +480,34 @@ export class PdfJsViewer {
...
-          ? this.queue.lastCompleted.mainFileId : undefined;
-      this.show(this.currentFileUrl, this.currentPage, true, lastCompletedMainFileId);
+  public showPage(page: number) {
+    if (this.currentFile === undefined){
+      return;
     }
+    if (page > this.currentFile.numPages){
+      page = this.currentFile.numPages;
+    }
+    let pageWrapper = this.loadedPages.find(x => x.id === page).div;
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]:
```
[bard@bardtop3 papeeria-pdfjs ((23f5c37...))]$ tsc
workspace/viewer/PdfJsViewer.ts(507,40): error TS2339: Property 'find' does not exist on type 'PDFPageViewStatic[]'.
```

Unfortunately find is not supported in IE so we'll have to resort on filter or map. This will also let us check if we really have found anything.

Also, there are no `PDFPageViewStatic.id` and `PDFPageViewStatic.div`. You pass `id` in the options to constructor, but it is not declared in `PdfPageViewStatic` interface (this is compiler's concern)  and it is not obvious that `id` becomes a property after passing it in the options (it is my concern). If it does appears as a property under the hood, please add appropriate declaration to the interface.

File: workspace/viewer/PdfJsViewer.ts [hunk 22]
```diff
@@ -515,49 +522,51 @@ export class PdfJsViewer {
...
+  getCurrentPage(): number | undefined {
+    return this.currentPage;
+  }
 
   getZoomScale(): number { return this.zoom.current(); }
 
   // Toolbar button handlers
   zoomIn = () => {
-    this.zoom.zoomIn() && this.openCurrentPage();
+    this.zoom.zoomIn() && this.showPage(this.currentPage);
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: Please fix zooming. Currently it changes zoom level, but the changes take affect only after subsequent calls to `resetCanvas` and `show`. When integrated into Papeeria it means that one has to recompile the file to change the zooming. I believe that it doesn't work properly in your simple UI either (you wrote that you added simple buttons as far as I remember. By the, please push that simple UI into the repository)

File: workspace/viewer/PdfJsViewer.ts (hunk 21 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts (hunk 22 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts [hunk 23]
```diff
@@ -550,11 +563,15 @@ export class PdfJsViewer {
...
   zoomIn = () => {
-    this.zoom.zoomIn() && this.showPage(this.currentPage);
+    if (this.currentFileUrl){
+      this.zoom.zoomIn() && this.showAll(this.currentFileUrl, true);  
+    }
   };
 
   zoomOut = () => {
-    this.zoom.zoomOut() && this.showPage(this.currentPage);
+    if (this.currentFileUrl){
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: pls whitespace before `{`
What IDE do you use? There must be code formating command. In IntelliJ it is Code >Reformat Code (it also works with the selection). Please consider using automated formatter in IDE because using me as formatter is kinda inefficient.

File: workspace/viewer/PdfJsViewer.ts [hunk 24]
```diff
@@ -550,11 +563,15 @@ export class PdfJsViewer {
 
   // Toolbar button handlers
   zoomIn = () => {
-    this.zoom.zoomIn() && this.showPage(this.currentPage);
+    if (this.currentFileUrl){
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: pls whitespace before `{`

File: workspace/viewer/PdfJsViewer.ts [hunk 25]
```diff
@@ -516,6 +519,16 @@ export class PdfJsViewer {
     }
   }
 
+  private findPageView(pageNumber: number) : PDFPageViewStatic | undefined {
+    if(!this.currentFile){
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: pls add whitespaces around `()`

File: workspace/viewer/PdfJsViewer.ts [hunk 26]
```diff
@@ -422,26 +422,29 @@ export class PdfJsViewer {
...
-        container: this.jqRoot.get(0),
-        id: pageNumber,
-        scale: scale,
-        defaultViewport: page.getViewport(1),
-        textLayerFactory: this.textLayerFactory
-      })
-      this.loadedPages.push(pageView);
+      if (this.currentTask.isResize) {
+        this.resetPage();
+        let pageView = this.findPageView(pageNumber);
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: This is unused code, isn't it? `let` is lexical scoped and is not visible outside of this `if` block.

File: workspace/viewer/PdfJsViewer.ts [hunk 27]
```diff
@@ -422,26 +422,29 @@ export class PdfJsViewer {
...
+        let pageView = new PDFPageView.PDFPageView({
+          container: this.jqRoot.get(0),
+          id: pageNumber,
+          scale: scale,
+          defaultViewport: page.getViewport(1),
+          textLayerFactory: this.textLayerFactory
+        })
+        this.loadedPages.push(pageView);
+      }
       pageView.update(scale);
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: And compiler agrees:
```
./build.sh 
workspace/viewer/PdfJsViewer.ts(442,7): error TS2304: Cannot find name 'pageView'.
workspace/viewer/PdfJsViewer.ts(443,7): error TS2304: Cannot find name 'pageView'.
workspace/viewer/PdfJsViewer.ts(455,7): error TS2304: Cannot find name 'pageView'.
```
Please consider running compiler on your code. Compiler and compile-time errors are the purpose of using TypeScript after all.

File: workspace/viewer/PdfJsViewer.ts [hunk 28]
```diff
@@ -515,49 +537,55 @@ export class PdfJsViewer {
...
+  getCurrentPage(): number | undefined {
+    return this.currentPage;
+  }
 
   getZoomScale(): number { return this.zoom.current(); }
 
   // Toolbar button handlers
   zoomIn = () => {
-    this.zoom.zoomIn() && this.openCurrentPage();
+    if (this.currentFileUrl){
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: This probably works for now but please add TODO to make it more delicate. It currently just reloads all the pages with new zoom scale which is probably not necessary.

**Comment by dbarashev** [comment]
For the record, I integrated this code with minor changes (compiler made happy) into a branch in the main Papeeria repository and it mostly works: scrolling, pagination and zoom in/ zoom out with buttons. Zoom presets (25%, 50%, etc) and zoom to fit width/page do not work. These actions call `zoomWidth`, `zoomPage` and `zoomPreset` methods. Please fix them as well.

File: workspace/viewer/PdfJsViewer.ts (hunk 26 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts (hunk 27 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts (hunk 25 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts (hunk 24 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts (hunk 23 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Thank you for patience. I've used Vim, but it's not adapted to find this type of errors, so I swaped back to phpstorm now.
Done.

File: workspace/viewer/PdfJsViewer.ts (hunk 28 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.
Now it updates the scale by calling `update(scale)` and draws(method `draw()`) page using new scale.

File: workspace/viewer/PdfJsViewer.ts [hunk 29]
```diff
@@ -440,8 +440,8 @@ export class PdfJsViewer {
       let scale = this.zoom.factor(page, this.jqRoot);
       let pageView;
       if (this.currentTask.isResize) {
-        this.resetPage();
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: Why did you remove this? It means that  you no longer call `zoom.onResize` which is probably doing some useful stuff (and it really does: it resets scale so that it was calculated the next time `factor` is called, so that page size could beset properly after viewer container resizing)

File: workspace/viewer/PdfJsViewer.ts [hunk 30]
```diff
@@ -451,9 +451,9 @@ export class PdfJsViewer {
           textLayerFactory: this.textLayerFactory
         })
         this.loadedPages.push(pageView);
+        pageView.update(scale);
+        pageView.setPdfPage(page);
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: What's the reason of this change? You no longer call `setPdfPage` in case of resize -- is it just not necessary?

File: workspace/viewer/PdfJsViewer.ts (hunk 29 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: I removed it because `setFitting()` function does exactly the same: changes `fittingScale` to `undefined` to recalculate size next time in cases of `ZoomingMode.FIT_PAGE`, `ZoomingMode.FIT_WIDTH`. Case of `ZoomingMode.PRESET` doesn't need `onResize` function to be called.

File: workspace/viewer/PdfJsViewer.ts (hunk 30 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: We find existing PDFPageView with page, which has been already set in case of resize. So we don't need to set it again and, yes, this function becomes not necessary.

File: workspace/viewer/PdfJsViewer.ts (hunk 29 above)
Comment (dbarashev) [reply workspace/viewer/PdfJsViewer.ts]: Why don't you call that function then? Before this change when flag `isResize` was set in the task it would resize page appropriately but now when this code is removed it won't.

Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Function `setFitting()` is called, but in `ZoomWidth()`, `ZoomPreset()`, `ZoomPage()`, which are executed before `openPage()` function, so it works fine.
I dont understand why we should evoke `resetPage()` if `setFitting()` does exactly the same?

Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Anyway, I returned function `resetPage()` back to keep going on.

File: workspace/viewer/PdfJsViewer.ts [hunk 31]
```diff
@@ -437,11 +437,13 @@ export class PdfJsViewer {
       if (!this.currentTask) {
         return false;
       }
+      if(this.currentTask.isResize){
```
Comment (dbarashev) [inline workspace/viewer/PdfJsViewer.ts]: spaces. around.

File: workspace/viewer/PdfJsViewer.ts (hunk 29 above)
Comment (dbarashev) [reply workspace/viewer/PdfJsViewer.ts]: Thanks for returning it back.

It is not the question of calling `resetPage` vs calling `setFitting`. It is the question of calling _anything_ which resets zooming level when we find a task with `resize == true`. Before these changes it was called and code in Papeeria which adds resize tasks expected such behavior. After these changes we stopped calling, and code in Papeeria got broken (for the user it meant that page won't resize properly when he resizes the viewer pane). 

But anyway we don't even  add such tasks on resize now.
Please replace onResize impementation with this one:
```
  onResize() {
    if (this.currentFile && this.currentFileUrl) {
      this.showAll(this.currentFileUrl, true);
    }
  }
```

File: workspace/viewer/PdfJsViewer.ts (hunk 31 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: Done.

File: workspace/viewer/PdfJsViewer.ts (hunk 29 above)
Comment (portuiu) [reply workspace/viewer/PdfJsViewer.ts]: It's clear. Thanks for explanation.
Done.

**Review by dbarashev — APPROVED** [review]
LGTM
I changes the base branch to `bardsoftware:project-smooth-scrolling` so you can just hit merge.

**Comment by dbarashev** [comment]
Oh, well, apparently there are merge conflicts which need to be resolved.

**Comment by dbarashev** [comment]
I just have merged these changes into `bardsoftware:project-smooth-scrolling` so this pull request can be closed now. 

The next thing is probably lazy rendering of pages: render page when it becomes visible in the viewport. Please fork the working branch from `project-smooth-scrolling` branch in `bardsoftware/papeeria-pdfjs` repository and commit changes to `bardsoftware/papeeria-pdfjs`, not to your own clone.
